File: c/binary-search/binary_search.c
#include "binary_search.h"


const int *binary_search(int value, const int *arr, size_t length) {

    if (length <= 0) {
        return NULL;
    }

    int low = 0;
    int high = length;

    while (low <= high) {
        int m = (low + high) / 2;

        if (value == arr[m]) {
            return &arr[m];
        }

        if (value > arr[m]) {
            low = m + 1;
        } else {
            high = m - 1;
        }
    }

    return NULL;
}

File: fortran/binary-search/binary_search.f90
module binary_search
  implicit none
contains

  function find(array, val) result(idx)
    integer, dimension(:), intent(in) :: array
    integer, intent(in) :: val
    integer :: idx, low, high, middle

    low = 1
    high = size(array)
    idx = -1

    do while (low <= high)
        middle = (low + high) / 2

        if (array(middle) .eq. val) then
            idx = middle
            exit
        end if

        if (array(middle) > val) then
            high = middle - 1
        else
            low = middle + 1
        end if

    end do

  end function

end module
